### PR TITLE
fix: [CDS-46113]: custom secret

### PR DESCRIPTION
**Pull request by shivamnegi94**
## Describe your changes
Currently We don't support secrets in perpetual task in NG and triggers create Perpetual task, because of the limitation customer can not use secrets in trigger.
Address the issue:

- we are now resolving the secret expression to ${ngSecretManager.obtain(””)} in NG Manager using CustomScriptSecretExpressionEvaluator class.
- ${ngSecretManager.obtain(””, secretfunctor)} resolving to ${secretDelegate.obtain("", token)} using NgCustomSecretExpressionEvaluator class.
- Added the code to substitute the expression in CustomArtifactService class delegate side

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/41186)

### CODE REVIEW

**Review by inderschera — APPROVED** [review]
> Reviewed 2 of 16 files at r1, 1 of 6 files at r2, 6 of 9 files at r4, 3 of 5 files at r5, 3 of 4 files at r6, 6 of 6 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41186)* status: 21 file(s) to review. 1 of 2 approvals obtained.

**Review by vivekkarnatiharness — APPROVED** [review]
> Reviewed 2 of 16 files at r1, 1 of 6 files at r2, 6 of 9 files at r4, 2 of 5 files at r5, 3 of 4 files at r6, 6 of 6 files at r8, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41186)* status: 1 file(s) to review. 1 of 2 approvals obtained.

File: 125-cd-nextgen/src/main/java/io/harness/cdng/artifact/mappers/ArtifactConfigToDelegateReqMapper.java [hunk 1]
```diff
@@ -201,39 +216,82 @@ public JenkinsArtifactDelegateRequest getJenkinsDelegateRequest(JenkinsArtifactC
...
+  }
+
+  public CustomScriptInlineSource getCustomScriptInlineSource(CustomArtifactConfig artifactConfig) {
+    CustomScriptInlineSource customScriptInlineSource = (CustomScriptInlineSource) artifactConfig.getScripts()
+                                                            .getFetchAllArtifacts()
+                                                            .getShellScriptBaseStepInfo()
+                                                            .getSource()
+                                                            .getSpec();
+    if (isNotEmpty(customScriptInlineSource.getScript().getValue())) {
+      if (isEmpty(artifactConfig.getScripts().getFetchAllArtifacts().getArtifactsArrayPath().getValue())) {
```
Comment (sribalij) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/artifact/mappers/ArtifactConfigToDelegateReqMapper.java:285]: If getArtifactsArraypath() is null won't getArtifactsArraypath().getValue() be null and give a NPE?

File: 125-cd-nextgen/src/main/java/io/harness/cdng/artifact/mappers/ArtifactConfigToDelegateReqMapper.java [hunk 2]
```diff
@@ -201,39 +216,82 @@ public JenkinsArtifactDelegateRequest getJenkinsDelegateRequest(JenkinsArtifactC
...
+        throw new InvalidArtifactServerException("Artifacts Array Path is missing", Level.ERROR, USER);
+      }
+      if (isEmpty(artifactConfig.getScripts().getFetchAllArtifacts().getVersionPath().getValue())) {
```
Comment (sribalij) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/artifact/mappers/ArtifactConfigToDelegateReqMapper.java:288]: Same with getVersionPath()?

File: 125-cd-nextgen/src/main/java/io/harness/cdng/artifact/utils/ArtifactStepHelper.java [hunk 3]
```diff
@@ -333,6 +342,15 @@ public ArtifactSourceDelegateRequest toSourceDelegateRequest(ArtifactConfig arti
             encryptedDataDetails, jenkinsArtifactConfig.getConnectorRef().getValue());
       case CUSTOM_ARTIFACT:
         CustomArtifactConfig customArtifactConfig = (CustomArtifactConfig) artifactConfig;
+        /*
+        We will be checking here if the request is from custom Triggers or from custom artifact source.
+        Version will be null in CustomArtifactConfig if the request is from custom trigger.
+        In Case of trigger, We will be resolving the expression in Ng Manager.
+         */
+        if (customArtifactConfig.isFromTrigger()) {
```
Comment (sribalij) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/artifact/utils/ArtifactStepHelper.java:350]: Any way we can combine the logic with custom trigger? How was it working earlier?

File: 125-cd-nextgen/src/main/java/io/harness/cdng/artifact/mappers/ArtifactConfigToDelegateReqMapper.java [hunk 4]
```diff
@@ -201,39 +216,82 @@ public JenkinsArtifactDelegateRequest getJenkinsDelegateRequest(JenkinsArtifactC
...
+      DelegateGrpcClientWrapper delegateGrpcClientWrapper, DelegateMetricsService delegateMetricsService,
+      SecretManagerClientService ngSecretService, ExceptionManager exceptionManager) {
+    NGAccess ngAccess = AmbianceUtils.getNgAccess(ambiance);
+    int secretFunctor = HashGenerator.generateIntegerHash();
+    CustomScriptInlineSource customScriptInlineSource = getCustomScriptInlineSource(artifactConfig);
+    String script = customScriptInlineSource.getScript().fetchFinalValue().toString();
+    script = resolveNGSecretExpression(script, secretFunctor);
+    NgCustomSecretExpressionEvaluator ngCustomSecretExpressionEvaluator = new NgCustomSecretExpressionEvaluator(script,
+        ngAccess.getAccountIdentifier(), ngAccess.getOrgIdentifier(), ngAccess.getProjectIdentifier(), secretFunctor,
+        delegateGrpcClientWrapper, delegateMetricsService, ngSecretService, exceptionManager);
```
Comment (sribalij) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/artifact/mappers/ArtifactConfigToDelegateReqMapper.java:250]: May not need exceptionManager passed to NgCustomSecretExpressionEvaluator

Comment (sribalij) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/artifact/mappers/ArtifactConfigToDelegateReqMapper.java:250]: Also delegateGrpcClientWrapperdelegateGrpcClientWrapper

File: 125-cd-nextgen/src/main/java/io/harness/cdng/artifact/mappers/ArtifactConfigToDelegateReqMapper.java [hunk 5]
```diff
@@ -201,39 +216,82 @@ public JenkinsArtifactDelegateRequest getJenkinsDelegateRequest(JenkinsArtifactC
...
         NGVariablesUtils.getStringMapVariables(artifactConfig.getInputs(), 0L), artifactConfig.getVersion().getValue(),
         ambiance != null ? AmbianceUtils.obtainCurrentRuntimeId(ambiance) : "", timeout,
         AmbianceUtils.getAccountId(ambiance));
   }
 
+  public CustomArtifactDelegateRequest getCustomDelegateRequest(CustomArtifactConfig artifactConfig, Ambiance ambiance,
+      DelegateGrpcClientWrapper delegateGrpcClientWrapper, DelegateMetricsService delegateMetricsService,
+      SecretManagerClientService ngSecretService, ExceptionManager exceptionManager) {
+    NGAccess ngAccess = AmbianceUtils.getNgAccess(ambiance);
+    int secretFunctor = HashGenerator.generateIntegerHash();
```
Comment (sribalij) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/artifact/mappers/ArtifactConfigToDelegateReqMapper.java:244]: Nit: Change the variable name to secretFuntor token?

**Review by sribalij — CHANGES_REQUESTED** [review]
few comments added

File: 125-cd-nextgen/src/main/java/io/harness/cdng/artifact/utils/ArtifactStepHelper.java (hunk 3 above)
Comment (shivamnegi94) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/artifact/utils/ArtifactStepHelper.java:350]: Same code is being used when we execute the pipeline to verify the build and secrets expression is taken care by delegate task

File: 125-cd-nextgen/src/main/java/io/harness/cdng/artifact/mappers/ArtifactConfigToDelegateReqMapper.java (hunk 5 above)
Comment (shivamnegi94) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/artifact/mappers/ArtifactConfigToDelegateReqMapper.java]: Resolved

**Comment by shivamnegi94** [comment]
trigger messagecheck

**Review by sribalij — APPROVED** [review]
> Reviewed 2 of 16 files at r1, 2 of 5 files at r5, 2 of 4 files at r6, 4 of 6 files at r8, 1 of 1 files at r9, 1 of 4 files at r10, 3 of 3 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41186)* status: 4 file(s) to review. 2. unresolved discussions. 1 of 2 approvals obtained.

**Review by vtxorxwitty — APPROVED** [review]
> Reviewed 2 of 16 files at r1, 6 of 9 files at r4, 2 of 5 files at r5, 2 of 4 files at r6, 4 of 6 files at r8, 1 of 1 files at r9, 1 of 4 files at r10, 3 of 3 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41186)* status: 2. unresolved discussions.

**Review by shivamnegi94 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41186)* status: 4. unresolved discussions. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/artifact/mappers/ArtifactConfigToDelegateReqMapper.java` line 250 at r9](https://reviewable.io/reviews/harness/harness-core/41186#-NIbC9aUFpHpfPQ3zvpo-r9-250:-NJ4JsHk4b7DsVeHLuz1:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/0fc9c760003ec35050d485e3a5dc842b263cf72e/125-cd-nextgen/src/main/java/io/harness/cdng/artifact/mappers/ArtifactConfigToDelegateReqMapper.java#L250)):*
<details><summary><i>Previously, sribalij (Sridhar) wrote…</i></summary><blockquote>

Also delegateGrpcClientWrapperdelegateGrpcClientWrapper
</blockquote></details>

Done.
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/artifact/mappers/ArtifactConfigToDelegateReqMapper.java` line 285 at r9](https://reviewable.io/reviews/harness/harness-core/41186#-NIbC9aUFpHpfPQ3zvpo-r9-285:-NJ4JtOA4QW-pl3Slyeq:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/0fc9c760003ec35050d485e3a5dc842b263cf72e/125-cd-nextgen/src/main/java/io/harness/cdng/artifact/mappers/ArtifactConfigToDelegateReqMapper.java#L285)):*
<details><summary><i>Previously, sribalij (Sridhar) wrote…</i></summary><blockquote>

If getArtifactsArraypath() is null won't getArtifactsArraypath().getValue() be null and give a NPE?
</blockquote></details>

Done.
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/artifact/mappers/ArtifactConfigToDelegateReqMapper.java` line 288 at r9](https://reviewable.io/reviews/harness/harness-core/41186#-NIbC9aUFpHpfPQ3zvpo-r9-288:-NJ4Ju7t7nfApzYCHnzX:bs7wumb) ([raw file](https://github.com/harness/harness-core/blob/0fc9c760003ec35050d485e3a5dc842b263cf72e/125-cd-nextgen/src/main/java/io/harness/cdng/artifact/mappers/ArtifactConfigToDelegateReqMapper.java#L288)):*
<details><summary><i>Previously, sribalij (Sridhar) wrote…</i></summary><blockquote>

Same with getVersionPath()?
</blockquote></details>

Done.

**Review by sribalij — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41186)* status: 1. unresolved discussions.

**Review by sribalij — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41186)* status: :shipit: complete! ✓

**Pull request merged**